### ISSUE TITLE
North star: log every judged memory operation as a training example

### ISSUE THREAD

**Issue by ronsse** [ronsse/trellis-ai#264]
**Strategic, not productionization** (deliberately no milestone). From the north star (\`docs/PRD.md\` §6, \`docs/design/plan-memory-lifecycle.md\` §0.1): a small local model eventually performs every judged memory operation. **Its training dataset accrues now or never** — every judged op already has the shape of a labeled example:

> (input context, decision, downstream outcome via feedback attribution)

**Design:**
- New EventLog event type (e.g. \`MEMORY_OP_JUDGED\`) emitted by every model-judged stage: extraction verdicts, reconcile-on-write ADD/UPDATE/SUPERSEDE/NOOP calls (#263), auto-capture distillations (#255), consolidation merge decisions. Payload: op type, model id, input digest (leak-safe — content refs, not raw prompt dumps of restricted items), decision, confidence.
- The **label arrives later**: feedback attribution joins the op to whether the memory it produced was ever served-and-useful. A periodic job (or trellis-evals tooling) materializes (input, decision, outcome) triples for export.
- Export tooling lives in **trellis-evals** (operator-owned), not core — core only emits events it already knows how to emit.

**Why now:** zero-cost bookkeeping on stages being built anyway (#255, #263); retrofitting labels onto un-logged history is impossible. When local models are ready to fine-tune, the corpus exists.

**Acceptance:** every judged stage emits the event (fixture per stage); a trellis-evals script materializes ≥1 labeled triple end-to-end from a synthetic session; DataClassification-restricted content never appears raw in payloads.

**Comment by ronsse** [comment]
## Scoping measurement — 2026-08-27

This is labelled `ready, mechanical`, which undersells how much of it already exists. Measured against `main` `9d3d59e` and the live event log, so nobody re-derives it:

**`memory_op.judged` is the second most common event type in production — 1,220 events**, behind only `mutation.executed` (1,662). The mechanism is not theoretical.

| `JudgedOpType` | emitter in `src/` | prod events | window |
|---|---|---|---|
| `CLASSIFICATION` | `classify/shadow.py:759` | **999** | 2026-08-23 → 08-24 |
| `DISTILLATION` | `trellis_workers/session_capture/distill.py:367` | **221** | 2026-08-24 → **08-27** |
| `RECONCILIATION` | `mcp/reconcile.py:370` | **0** | — |
| `EXTRACTION` | **none** | 0 | — |
| `CURATION` | **none** | 0 | — |

So the remaining work is **two stages with no emitter at all** (`EXTRACTION`, `CURATION`), not five.

### Three things the table hides, each worth knowing before starting

1. **`RECONCILIATION` has an emitter but has never fired.** Not a defect — `reconcile_on_write` is `false` in production (confirmed via `admin write-config`). The emitter described in `mcp/reconcile.py` as "the first `MEMORY_OP_JUDGED` emitter" has produced zero events because the feature that calls it is switched off. Anyone measuring coverage by counting emitters would score this stage as done.

2. **`CLASSIFICATION`'s 999 events are one batch pass, not a live stream.** They span two days and stop. That is correct by design — shadow mode is deliberately a batch pass, since an LLM call has no business in the inline write path — but it means classification training pairs accrue only when someone runs `trellis classify shadow` by hand. `DISTILLATION` is the only stage currently producing events continuously (nightly session-capture cron at 03:00).

3. **Volume is dominated by whichever stage last had a batch run**, so a single global count of `memory_op.judged` is a poor progress metric for this issue. Break it down by `op_type`, as above.

### Suggestion

Worth splitting: `EXTRACTION` and `CURATION` emitters are the genuinely mechanical part and could be one PR. Whether `RECONCILIATION` counts as "done" is a separate judgement about what this issue is measuring — emitters written, or training pairs actually accruing. Those give different answers, and the second is the one the issue title ("log every judged memory operation as a training example") is asking for.

**Comment by ronsse** [comment]
Implementation plan (premise re-verified against `f9ff32c`, scope cut, mutants and gate risks named): https://github.com/ronsse/trellis-ai/blob/handoff/issue-sweep-2026-09-04/docs/handoff/2026-09-04/plans/264.md — part of the 2026-09-04 review-and-simplify sweep; orchestration notes in https://github.com/ronsse/trellis-ai/blob/handoff/issue-sweep-2026-09-04/docs/handoff/2026-09-04/HANDOFF.md. The branch is a handoff record, not for merge.

**Comment by ronsse** [comment]
**Remainder statement (not a close), plus a corrected estimate for PR-B.**

PR #540 shipped the classify-layer instance. The remainder is **PR-B: the derived roster of judged sites**, and the sequencing note attached to it in the 2026-09-04 corpus — *"after #514 when `generate_call_sites` exists"* — was **wrong about which helper does the work**. Correcting it here so nobody re-derives the wrong estimate.

**`generate_call_sites` does not generalize.** It exists at `tests/ast_rules.py:405` and is exported, but it is hard-wired to LLM `generate(messages=...)`:

```python
if any(keyword.arg == "messages" or keyword.arg is None for keyword in node.keywords)
```

There is no subject parameter. PR-B needs its own ~20-line predicate over `emit_memory_op_judged` call sites; it cannot call this one.

**But the part that is easy to get wrong is fully reusable, which is the more useful half of this correction.** The three vacuity guards are already generic:

- `assert_hand_read_floor(population, floor, *, subject, hint, sole_site_reason)` (`:1495`)
- `assert_scan_is_not_vacuous(shipped_predicate: Callable[[Path], Iterable[int]], ..., subject, tmp_path, live_population, floor)` (`:1640`) — it takes the **shipped predicate as an argument**, so it runs whatever scanner PR-B writes against the evasion tree
- `render_evasion_corpus(..., *, subject, args="", kwarg="gate", wrap="{call}")` (`:1105`) — parameterised by subject and keyword, so the `MARKER` / `DECOY_MARKER` roster and its five axes (`spelling`, `binding`, `placement`, `judgement`, `discovery`) apply unchanged

plus `iter_modules`, `construction_names`, `calls_to_any`, `marker_lines`, `decoy_lines`.

**Net:** PR-B is *not* "call an existing helper", and it is *not* "build a derived rule from scratch" either. It is one new predicate wired into machinery that already exists and is already proved — including `assert_scan_is_not_vacuous`, whose own cross-check was proved by re-introducing #457's under-collection bug and asserting it fails naming the dropped line. Reuse those; do not write a second set.

**Trap carried forward from #540's operational note (ledger A-3):** `EXTRACTION` judged-event counts are **zero in production** while `TRELLIS_ENABLE_MEMORY_EXTRACTION` is off. A rule that divides by observed emissions measures the flag, not the roster. Derive the floor from the source tree, hand-read, exactly as `test_governed_write_rule.py:126-129` does — *"never from `len()` of the scan they bound"*.
